src/main/scala/org/hammerlab/guacamole/reads/Read.scala

```diff
@@ -294,109 +261,56 @@ object Read extends Logging {
-
-      SamReaderFactory.setDefaultValidationStringency(ValidationStringency.SILENT)
-      val factory = SamReaderFactory.makeDefault
-      val reader = factory.open(new File(path.toUri.getPath))
-      val samSequenceDictionary = reader.getFileHeader.getSequenceDictionary
-      val sequenceDictionary = SequenceDictionary.fromSAMSequenceDictionary(samSequenceDictionary)
-      val loci = filters.overlapsLoci.map(_.result(contigLengths(sequenceDictionary)))
-      val recordIterator: SAMRecordIterator = if (filters.overlapsLoci.nonEmpty && reader.hasIndex) {
-        Common.progress("Using samtools with BAM index to read: %s".format(filename))
-        requiresFilteringByLocus = false
-        val queryIntervals = loci.get.contigs.flatMap(contig => {
-          val contigIndex = samSequenceDictionary.getSequenceIndex(contig)
-          loci.get.onContig(contig).ranges().map(range =>
-            new QueryInterval(contigIndex,
-              range.start.toInt + 1, // convert 0-indexed inclusive to 1-indexed inclusive
-              range.end.toInt)) // "convert" 0-indexed exclusive to 1-indexed inclusive, which is a no-op)
-        })
-        val optimizedQueryIntervals = QueryInterval.optimizeIntervals(queryIntervals.toArray)
-        reader.query(optimizedQueryIntervals, false) // Note: this can return unmapped reads, which we filter below.
-      } else {
-        val skippedReason = if (reader.hasIndex)
-          "(index is available but not needed)"
-        else
-          "(index unavailable)"
-        Common.progress("Using samtools without BAM index %s to read: %s".format(skippedReason, filename))
-        reader.iterator
-      }
-      val reads = JavaConversions.asScalaIterator(recordIterator).flatMap(record => {
-        // Optimization: some of the filters are easy to run on the raw SamRecord, so we avoid making a Read.
-        if ((filters.overlapsLoci.nonEmpty && record.getReadUnmappedFlag) ||
-          (requiresFilteringByLocus &&
-            !loci.get.onContig(record.getContig).intersects(record.getStart - 1, record.getEnd)) ||
-            (filters.nonDuplicate && record.getDuplicateReadFlag) ||
-            (filters.passedVendorQualityChecks && record.getReadFailsVendorQualityCheckFlag) ||
-            (filters.isPaired && !record.getReadPairedFlag)) {
-          None
-        } else {
-          val read = fromSAMRecord(record)
-          assert(filters.overlapsLoci.isEmpty || read.isMapped)
-          Some(read)
-        }
-      })
-      (sc.parallelize(reads.toSeq), sequenceDictionary)
+
+    val conf = sc.hadoopConfiguration
+    val samHeader = SAMHeaderReader.readSAMHeaderFrom(path, conf)
+    val sequenceDictionary = SequenceDictionary.fromSAMHeader(samHeader)
+
+    if (filters.overlapsLoci.isEmpty || filename.endsWith(".bam")) {
+      val lociString = filters.overlapsLoci.map(_.result(contigLengths(sequenceDictionary)).toString)
+      lociString.foreach(conf.set(BAMInputFormat.INTERVALS_PROPERTY,_))
```

Comment:
so our loci-string syntax is interoperable with hadoop-bam's? any 0- versus 1-based indexing gotchas?